<a href="https://colab.research.google.com/github/Krzesimir13/Data-Science/blob/main/141_150_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Pandas

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 141](#1)
* [Ćwiczenie 142](#2)
* [Ćwiczenie 143](#3)
* [Ćwiczenie 144](#4)
* [Ćwiczenie 145](#5)
* [Ćwiczenie 146](#6)
* [Ćwiczenie 147](#7)
* [Ćwiczenie 148](#8)
* [Ćwiczenie 149](#9)
* [Ćwiczenie 150](#10)

### <a name='0'></a>Import biblioteki

In [1]:
import numpy as np
import pandas as pd

np.random.seed(42)
np.__version__

'2.0.2'

### <a name='1'></a> Ćwiczenie 141
Podane są dwa obiekty typu _Series_:

```
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))
```
Połącz te dwa obiekty w jeden obiekt typu _DataFrame_ (jako dwie kolumny) i przypisz do zmiennej _df_. Nadaj nazwy kolumn odpowiednio _col1_ oraz _col2_.


In [2]:
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))

df = pd.concat([s1, s2],keys=['col1', 'col2'], axis=1)
df

col1      col2
0   0.374540 -1.012831
1   0.950714  0.314247
2   0.731994 -0.908024
3   0.598658 -1.412304
4   0.156019  1.465649
5   0.155995 -0.225776
6   0.058084  0.067528
7   0.866176 -1.424748
8   0.601115 -0.544383
9   0.708073  0.110923
10  0.020584 -1.150994
11  0.969910  0.375698
12  0.832443 -0.600639
13  0.212339 -0.291694
14  0.181825 -0.601707
15  0.183405  1.852278
16  0.304242 -0.013497
17  0.524756 -1.057711
18  0.431945  0.822545
19  0.291229 -1.220844

### <a name='2'></a> Ćwiczenie 142
Wytnij z obiektu _df_ wiersze, dla których wartość w kolumnie _col2_ jest pomiędzy 0.0 a 1.0 (włącznie).

In [3]:
df.loc[df['col2'].between(0.0, 1.0, inclusive='both')]

col1      col2
1   0.950714  0.314247
6   0.058084  0.067528
9   0.708073  0.110923
11  0.969910  0.375698
18  0.431945  0.822545

### <a name='3'></a> Ćwiczenie 143
Przypisz nową kolumnę o nazwie _col3_, która przyjmie wartość 1 w momencie gdy w kolumnie _col2_ jest wartość nieujemna i przeciwnie -1.

In [11]:
df['col3'] = 1
df.loc[df['col2'] < 0, 'col3'] = -1
df.col3

0    -1
1     1
2    -1
3    -1
4     1
5    -1
6     1
7    -1
8    -1
9     1
10   -1
11    1
12   -1
13   -1
14   -1
15    1
16   -1
17   -1
18    1
19   -1
Name: col3, dtype: int64

### <a name='4'></a> Ćwiczenie 144
Przypisz nową kolumnę o nazwie _col4_, która przytnie wartości z kolumny _col2_ do przedziału $[-1.0, 1.0]$.
Innymi słowy, dla wartości poniżej -1.0 należy wstawić wartość -1.0, dla wartości powyżej 1.0 wartość 1.0.


In [12]:
df['col4'] = df['col2']
df.loc[df['col2'] < -1.0, 'col4'] = -1.0
df.loc[df['col2'] > 1.0, 'col4'] = 1.0
df.col4

0    -1.000000
1     0.314247
2    -0.908024
3    -1.000000
4     1.000000
5    -0.225776
6     0.067528
7    -1.000000
8    -0.544383
9     0.110923
10   -1.000000
11    0.375698
12   -0.600639
13   -0.291694
14   -0.601707
15    1.000000
16   -0.013497
17   -1.000000
18    0.822545
19   -1.000000
Name: col4, dtype: float64

### <a name='5'></a> Ćwiczenie 145
Znajdź 5 największych wartości dla kolumny _col2_.

In [ ]:
# tutaj wpisz rozwiązanie

Znajdź 5 najmniejszych wartości dla kolumny _col2_.

In [13]:
df['col2'].sort_values(ascending=True).head(5)

7    -1.424748
3    -1.412304
19   -1.220844
10   -1.150994
17   -1.057711
Name: col2, dtype: float64

### <a name='6'></a> Ćwiczenie 146
Wyznacz skumulowaną sumę dla każdej kolumny.

__Wskazówka:__ Użyj metody _pd.DataFrame.cumsum()_.

In [17]:
df.cumsum()

col1      col2      col3      col4
0   0.374540 -0.638291 -1.638291 -2.638291
1   0.950714  1.264962  2.264962  2.579209
2   0.731994 -0.176030 -1.176030 -2.084054
3   0.598658 -0.813645 -1.813645 -2.813645
4   0.156019  1.621667  2.621667  3.621667
5   0.155995 -0.069782 -1.069782 -1.295558
6   0.058084  0.125612  1.125612  1.193140
7   0.866176 -0.558572 -1.558572 -2.558572
8   0.601115  0.056732 -0.943268 -1.487650
9   0.708073  0.818995  1.818995  1.929918
10  0.020584 -1.130409 -2.130409 -3.130409
11  0.969910  1.345608  2.345608  2.721306
12  0.832443  0.231804 -0.768196 -1.368835
13  0.212339 -0.079355 -1.079355 -1.371048
14  0.181825 -0.419882 -1.419882 -2.021588
15  0.183405  2.035683  3.035683  4.035683
16  0.304242  0.290745 -0.709255 -0.722752
17  0.524756 -0.532954 -1.532954 -2.532954
18  0.431945  1.254490  2.254490  3.077035
19  0.291229 -0.929615 -1.929615 -2.929615

### <a name='7'></a> Ćwiczenie 147
Wyznacz medianę dla zmiennej _col2_ (inaczej kwantyl rzędu 0.5).

In [19]:
df['col2'].median()

-0.4180382371592297

### <a name='8'></a> Ćwiczenie 148
Wytnij wiersze, dla których zmienna _col2_ przyjmuje wartości większe od 0.0.

In [20]:
df.loc[df['col2'] > 0.0]

col1      col2  col3      col4
1   0.950714  0.314247     1  0.314247
4   0.156019  1.465649     1  1.000000
6   0.058084  0.067528     1  0.067528
9   0.708073  0.110923     1  0.110923
11  0.969910  0.375698     1  0.375698
15  0.183405  1.852278     1  1.000000
18  0.431945  0.822545     1  0.822545

### <a name='9'></a> Ćwiczenie 149
Wytnij 5 pierwszych wierszy obiektu _df_ i przekształć je do słownika.

In [22]:
df.iloc[5].to_dict()

{'col1': 0.15599452033620265,
 'col2': -0.22577630048653566,
 'col3': -1.0,
 'col4': -0.22577630048653566}

### <a name='10'></a> Ćwiczenie 150
Wytnij 5 pierwszych wierszy obiektu _df_, przekształć je do formatu Markdown i przypisz do zmiennej _df_markdown_.

In [23]:
df_markdown = df.iloc[5].to_markdown()

Wydrukuj do konsoli zawartość zmiennej _df_markdown_.

In [25]:
print(df_markdown)

|      |         5 |
|:-----|----------:|
| col1 |  0.155995 |
| col2 | -0.225776 |
| col3 | -1        |
| col4 | -0.225776 |


Skopiuj wynik uruchomienia powyższej komórki i wklej poniżej:

|      |         5 |
|:-----|----------:|
| col1 |  0.155995 |
| col2 | -0.225776 |
| col3 | -1        |
| col4 | -0.225776 |